tmt/beakerlib.py

```diff
@@ -191,12 +200,16 @@ def fetch(self) -> None:
             if library.url != self.url:
                 # tmt guessed url so try if repo exists
                 if self.format == 'rpm':
+                    if self.url in self._nonexistent_url:
+                        self.parent.debug(f"Already know '{self.url}' is not found.")
+                        raise LibraryError
                     with TemporaryDirectory() as tmp:
                         try:
                             tmt.utils.git_clone(str(self.url), Path(tmp), self.parent,
                                                 env={"GIT_ASKPASS": "echo"}, shallow=True)
                         except tmt.utils.RunError:
                             self.parent.debug(f"Repository '{self.url}' not found.")
+                            self._nonexistent_url.add(str(self.url))
```

Comment:
Why is `str()` needed? IIUIC, `self.url` is `Optional[str]`, therefore it's either:

* `None`, but then we're calling `tmt.utils.git_clone('None', ...) on line 208, which hardly seems correct, or
* it's a string already, then we don't need `str()` at all, here and in the `git_clone()` call as well, or
* it's a different type, but then the annotation is wrong, or the type should never got here.

It seems like `str()` should not be used here at all, the possible `str(None)` conversion does not seem right. We lack either the `self.url is None` branch, somewhere, no idea where, I don't understand this bit of code well enough, or `self.url` simply cannot be `None`, and we should add `assert self.url is not None  # narrow type` to signal that to readers & drop `str()` calls.

I *think* some handling is happening below:

```python
library = self._library_cache[str(self.repo)]
```

If `self.repo` is `None`, `KeyError` might be raised, and the `except KeyError` branch does check whether `self.url` is a string or not... Making assumptions about type of `self.url` is hard for me :/

Reply:
Fixed in e65427a